ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,189 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
                           IN CONST GUID *AliasInterfaceClassGuid,
                           OUT PUNICODE_STRING AliasSymbolicLinkName)
 {
-    return STATUS_NOT_IMPLEMENTED;
+    UNICODE_STRING BaseKeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Control\\Class\\");
+    UNICODE_STRING AliasKeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\");
+    UNICODE_STRING bipName, InLinkName;
+    UNICODE_STRING GuidString;
+    HANDLE InterfaceKey = NULL;
+    PKEY_FULL_INFORMATION fip = NULL;
+    PKEY_BASIC_INFORMATION bip = NULL;
+    PWCHAR pdest;
+    ULONG i, Size;
+    LONG ret;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    BaseKeyName.Length = 0;
+    BaseKeyName.MaximumLength += GuidString.Length;
+    BaseKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                               BaseKeyName.MaximumLength,
+                                               TAG_IO);
+    if (!BaseKeyName.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    /* Concatenate alias GUID string with base key */
+    Status = RtlAppendUnicodeStringToString(&BaseKeyName, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Open base registry key */
+    Status = IoOpenDeviceInterfaceRegistryKey(&BaseKeyName,
+                                              GENERIC_READ,
+                                              &InterfaceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IoOpenDeviceInterfaceRegistryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        NULL,
+                        0,
+                        &Size);
+    if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    fip = ExAllocatePoolWithTag(NonPagedPool,
```

Comment:
Using `PagedPool` for "fip" should be fine, since your function runs in PAGED_MODE and this "fip" is used locally there only.

Reply:
Fixed.